client/sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java

```diff
@@ -51,103 +52,84 @@
     private static final Log logger = LogFactory.getLog(Sniffer.class);
     private static final String SNIFFER_THREAD_NAME = "es_rest_client_sniffer";
 
-    private final Task task;
+    private final HostsSniffer hostsSniffer;
+    private final RestClient restClient;
+
+    private final long sniffIntervalMillis;
+    private final long sniffAfterFailureDelayMillis;
+    private final Scheduler scheduler;
+
+    private final AtomicReference<Future> nextTask = new AtomicReference<>();
 
     Sniffer(RestClient restClient, HostsSniffer hostsSniffer, long sniffInterval, long sniffAfterFailureDelay) {
-        this.task = new Task(hostsSniffer, restClient, sniffInterval, sniffAfterFailureDelay);
+        this(restClient, hostsSniffer, new DefaultScheduler(), sniffInterval, sniffAfterFailureDelay);
+    }
+
+    Sniffer(RestClient restClient, HostsSniffer hostsSniffer, Scheduler scheduler,  long sniffInterval, long sniffAfterFailureDelay) {
+        this.hostsSniffer = hostsSniffer;
+        this.restClient = restClient;
+        this.sniffIntervalMillis = sniffInterval;
+        this.sniffAfterFailureDelayMillis = sniffAfterFailureDelay;
+        this.scheduler = scheduler;
+        //first sniffing round is immediately executed, next one will be executed depending on the configured sniff interval
+        scheduleNextRound(0L, sniffIntervalMillis, false);
     }
 
     /**
-     * Triggers a new sniffing round and explicitly takes out the failed host provided as argument
+     * Triggers a new immediate sniffing round, which will schedule a new round in sniffAfterFailureDelayMillis ms
      */
-    public void sniffOnFailure(HttpHost failedHost) {
-        this.task.sniffOnFailure(failedHost);
+    public final void sniffOnFailure() {
+        scheduleNextRound(0L, sniffAfterFailureDelayMillis, true);
     }
 
-    @Override
-    public void close() throws IOException {
-        task.shutdown();
+    private void scheduleNextRound(long delay, long nextDelay, boolean mustCancelNextRound) {
+        Task task = new Task(nextDelay);
+        Future<?> nextFuture = task.schedule(delay);
+        Future<?> previousFuture = nextTask.getAndSet(nextFuture);
+        if (mustCancelNextRound) {
+            previousFuture.cancel(false);
```

Comment:
When I was playing with cancellation as part of reindex I found that canceling a Runnable was sort of "best effort". If you make a test that calls `sniffOnFailure` a bunch of time really fast together I'll bet you get multiple rounds of sniffing in parallel.

Reply:
I guess it depends on whether the runnable has already started or not? That is what I've seen, but hard to test in real-life though from a unit test...